jolt-core/src/subprotocols/ra_virtual.rs

```diff
@@ -0,0 +1,204 @@
+use crate::{
+    field::JoltField,
+    poly::{
+        eq_poly::EqPolynomial,
+        multilinear_polynomial::{BindingOrder, MultilinearPolynomial, PolynomialBinding},
+    },
+    subprotocols::sumcheck::{BatchableSumcheckInstance, SumcheckInstanceProof},
+    utils::{math::Math, transcript::Transcript},
+};
+
+use crate::poly::multilinear_polynomial::PolynomialEvaluation;
+use rayon::iter::IndexedParallelIterator;
+use rayon::iter::{IntoParallelRefIterator, IntoParallelRefMutIterator, ParallelIterator};
+pub struct RAProof<F: JoltField, ProofTranscript: Transcript> {
+    pub ra_i_claims: Vec<F>,
+    pub sumcheck_proof: SumcheckInstanceProof<F, ProofTranscript>,
+}
+
+/// Virtual RA sumcheck for d-way chunked addresses
+pub struct RASumcheck<F: JoltField, const D: usize> {
+    /// ra(r_cycle, r_address)
+    ra_claim: F,
+    /// ra polynomials for each chunk
+    ra_i_polys: [MultilinearPolynomial<F>; D],
+    /// Eq polynomial as a multilinear polynomial
+    eq_poly: MultilinearPolynomial<F>,
+    /// Random point r_cycle
+    r_cycle: Vec<F>,
+    /// r_address pre-chunked
+    r_address: Vec<F>,
+    /// Random points r_address^(i) for each chunk
+    r_address_chunks: [Vec<F>; D],
+    /// ra_i_ claims to be proved via evaluation proof
+    ra_i_claims: Option<[F; D]>,
+    /// Length of the trace
+    T: usize,
+}
+
+impl<F: JoltField, const D: usize> RASumcheck<F, D> {
+    pub fn new(
+        ra_claim: F,
+        mut ra_i_polys: [MultilinearPolynomial<F>; D],
+        r_cycle: Vec<F>,
+        r_address: Vec<F>,
+        T: usize,
+    ) -> Self {
+        assert_eq!(
+            r_address.len() % D,
+            0,
+            "r_address length must be divisible by D"
+        );
+
+        let chunk_size = r_address.len() / D;
+        let mut r_address_chunks_vec = Vec::with_capacity(D);
+
+        for i in 0..D {
+            let start = i * chunk_size;
+            let end = (i + 1) * chunk_size;
+            r_address_chunks_vec.push(r_address[start..end].to_vec());
+        }
+
+        let r_address_chunks: [Vec<F>; D] = r_address_chunks_vec
+            .try_into()
+            .expect("Failed to convert Vec to array");
+
+
+        let eq_evals = EqPolynomial::evals(&r_cycle);
+
+        let eq_poly = MultilinearPolynomial::from(eq_evals);
+
+        // We do pre-binding. The sumcheck is in variable j hence we can pre-bind the r_address chunks:
+        for (poly, chunk) in ra_i_polys.iter_mut().zip(r_address_chunks.iter()) {
+            for &r_value in chunk.iter() {
+                poly.bind_parallel(r_value, BindingOrder::LowToHigh);
+            }
+        }
```

Comment:
Instead of "pre-binding", which requires that we fully materialize each ra_i polynomial (which have length K^(1/d) * T), we should do the following:
- Compute the table of evals for EQ(k_i, r_address_chunks[i]) for each `i in 0..D`. Each table is size K^(1/d), which is small and quick to compute. 
- Then compute each pre-bound ra_i polynomial (i.e. ra_i(r_address_chunks[i], i)) with a single pass over the original addresses. We sort of do this in prefix-suffix Shout [here](https://github.com/a16z/jolt/blob/feat/twist-shout-integration/jolt-core/src/subprotocols/sparse_dense_shout.rs#L503-L510) –– `v` is the EQ table described above, just constructed incrementally over the course of the prefix-suffix phase. This takes time O(d * T)